File: python-skeleton/game_state.py
"""
Tankie – local game state tracker.

Keep this updated as events arrive from the server so your strategy
can make informed decisions without blind guessing.
"""

from dataclasses import dataclass, field


@dataclass
class TankState:
    tank_id: str
    pos_x: float = 0.0
    pos_y: float = 0.0
    grid_x: int = 0
    grid_y: int = 0
    turret_degrees: float = 0.0
    alive: bool = True


@dataclass
class GameState:
    my_tank_id: str
    tanks: dict[str, TankState] = field(default_factory=dict)
    on_turn: str = ""
    game_started: bool = False
    game_over: bool = False
    winner: str | None = None

    # Map data – populated by the "map" event at game start
    grid_width: int = 0
    grid_height: int = 0
    grid_size: int = 128
    walls: set[tuple[int, int, str]] = field(default_factory=set)

    # ------------------------------------------------------------------ #
    # Event handlers – call these from your main loop when an event arrives
    # ------------------------------------------------------------------ #

    def handle_player_joined(self, tank_id: str) -> None:
        if tank_id not in self.tanks:
            self.tanks[tank_id] = TankState(tank_id=tank_id)

    def handle_map(self, data: dict) -> None:
        self.game_over = False
        self.game_started = False
        self.tanks.clear()

        self.grid_width = data["gridWidth"]
        self.grid_height = data["gridHeight"]
        self.grid_size = data["gridSize"]
        self.walls = {
            (w["x"], w["y"], w["orientation"]) for w in data["walls"]
        }
        for p in data["players"]:
            tid = p["tankId"]
            if tid not in self.tanks:
                self.tanks[tid] = TankState(tank_id=tid)
            self.tanks[tid].grid_x = p["x"]
            self.tanks[tid].grid_y = p["y"]

    def handle_game_started(self, on_turn: str) -> None:
        self.game_started = True
        self.on_turn = on_turn

    def handle_turn_changed(self, next_turn: str) -> None:
        self.on_turn = next_turn

    def handle_moved(self, tank_id: str, direction: str, pos_x: float, pos_y: float) -> None:
        if tank_id in self.tanks:
            tank = self.tanks[tank_id]
            tank.pos_x = pos_x
            tank.pos_y = pos_y
            tank.grid_x = int(pos_x) // self.grid_size
            tank.grid_y = int(pos_y) // self.grid_size

    def handle_rotated(self, tank_id: str, degrees: float) -> None:
        if tank_id in self.tanks:
            self.tanks[tank_id].turret_degrees = degrees

    def handle_game_over(self, winner: str) -> None:
        self.game_over = True
        self.winner = winner

    def handle_turn_timeout(self, tank_id: str) -> None:
        pass  # Turn skipped; server will send turn_changed next

    def handle_turn_disqualified(self, tank_id: str) -> None:
        if tank_id in self.tanks:
            self.tanks[tank_id].alive = False

    # ------------------------------------------------------------------ #
    # Convenience helpers
    # ------------------------------------------------------------------ #

    @property
    def is_my_turn(self) -> bool:
        return self.game_started and self.on_turn == self.my_tank_id

    def my_tank(self) -> TankState | None:
        return self.tanks.get(self.my_tank_id)

    def opponents(self) -> list[TankState]:
        return [t for tid, t in self.tanks.items() if tid != self.my_tank_id and t.alive]

    def has_wall(self, x: int, y: int, orientation: str) -> bool:
        """Check if a wall exists at (x, y) with the given orientation."""
        return (x, y, orientation) in self.walls

File: csharp-skeleton/TankieClient/GameState.cs
using System.Collections.Generic;
using System.Linq;
using System.Text.Json.Nodes;

namespace TankieClient;

/// <summary>
/// Snapshot of a single tank as reported by the server.
/// </summary>
public class TankState
{
    public string TankId        { get; init; } = "";
    public float  PosX          { get; set; }
    public float  PosY          { get; set; }
    public int    GridX         { get; set; }
    public int    GridY         { get; set; }
    public float  TurretDegrees { get; set; }
    public bool   Alive         { get; set; } = true;
}

/// <summary>
/// Local mirror of the game state, updated from incoming server events.
/// </summary>
public class GameState
{
    public string MyTankId    { get; }
    public string OnTurn      { get; private set; } = "";
    public bool   GameStarted { get; private set; }
    public bool   GameOver    { get; private set; }
    public string? Winner     { get; private set; }

    // Map data – populated by the "map" event at game start
    public int GridWidth  { get; private set; }
    public int GridHeight { get; private set; }
    public int GridSize   { get; private set; } = 128;

    private readonly Dictionary<string, TankState> _tanks = new();
    private readonly HashSet<(int X, int Y, string Orientation)> _walls = new();

    public GameState(string myTankId)
    {
        MyTankId = myTankId;
    }

    // ------------------------------------------------------------------ //
    // Event handlers – call these from the main loop when an event arrives
    // ------------------------------------------------------------------ //

    public void HandlePlayerJoined(string tankId)
    {
        if (!_tanks.ContainsKey(tankId))
            _tanks[tankId] = new TankState { TankId = tankId };
    }

    public void HandleMap(JsonObject data)
    {
        GameOver    = false;
        GameStarted = false;
        _tanks.Clear();

        GridWidth  = data["gridWidth"]!.GetValue<int>();
        GridHeight = data["gridHeight"]!.GetValue<int>();
        GridSize   = data["gridSize"]!.GetValue<int>();

        _walls.Clear();
        foreach (var wallNode in data["walls"]!.AsArray())
        {
            var w = wallNode!.AsObject();
            _walls.Add((
                w["x"]!.GetValue<int>(),
                w["y"]!.GetValue<int>(),
                w["orientation"]!.GetValue<string>()
            ));
        }

        foreach (var playerNode in data["players"]!.AsArray())
        {
            var p      = playerNode!.AsObject();
            var tankId = p["tankId"]!.GetValue<string>();
            if (!_tanks.ContainsKey(tankId))
                _tanks[tankId] = new TankState { TankId = tankId };
            _tanks[tankId].GridX = p["x"]!.GetValue<int>();
            _tanks[tankId].GridY = p["y"]!.GetValue<int>();
        }
    }

    public void HandleGameStarted(string onTurn)
    {
        GameStarted = true;
        OnTurn = onTurn;
    }

    public void HandleTurnChanged(string nextTurn)
    {
        OnTurn = nextTurn;
    }

    public void HandleMoved(string tankId, string direction, float posX, float posY)
    {
        if (_tanks.TryGetValue(tankId, out var tank))
        {
            tank.PosX  = posX;
            tank.PosY  = posY;
            tank.GridX = (int)posX / GridSize;
            tank.GridY = (int)posY / GridSize;
        }
    }

    public void HandleRotated(string tankId, float degrees)
    {
        if (_tanks.TryGetValue(tankId, out var tank))
            tank.TurretDegrees = degrees;
    }

    public void HandleGameOver(string winner)
    {
        GameOver = true;
        Winner   = winner;
    }

    // ------------------------------------------------------------------ //
    // Convenience helpers
    // ------------------------------------------------------------------ //

    /// <summary>True when the game is running and it is our turn.</summary>
    public bool IsMyTurn => GameStarted && OnTurn == MyTankId;

    public TankState? MyTank() => _tanks.GetValueOrDefault(MyTankId);

    public IEnumerable<TankState> Opponents() =>
        _tanks.Values.Where(t => t.TankId != MyTankId && t.Alive);

    /// <summary>Check if a wall exists at (x, y) with the given orientation.</summary>
    public bool HasWall(int x, int y, string orientation) =>
        _walls.Contains((x, y, orientation));
}
